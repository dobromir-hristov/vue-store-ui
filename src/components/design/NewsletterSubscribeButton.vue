<template>
  <newsletter-subscribe-provider
    v-slot="{ toggleSubscription, loading, disabled, isSubscribed }"
    newsletter-id="main_newsletter"
  >
    <v-button
      :loading="loading"
      :disabled="disabled"
      size="large"
      type="alternative"
      :outlined="!isSubscribed"
      @click="toggleSubscription"
    >
      <span v-if="isSubscribed">Unsubscribe</span>
      <span v-else>Subscribe</span>
    </v-button>
  </newsletter-subscribe-provider>
</template>

<script>
import NewsletterSubscribeProvider from '../functional/NewsletterSubscribeProvider'

/**
 * Displays a subscribe/unsub newsletter button
 * @see module:NewsletterSubscribeProvider
 * @module NewsletterSubscribeButton
 */
export default {
  name: 'NewsletterSubscribeButton',
  components: { NewsletterSubscribeProvider }
}
</script>
