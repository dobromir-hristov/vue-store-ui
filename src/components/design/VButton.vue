<template>
  <button
    :class="[`v-button--${color}`, { 'is-outlined': outlined }]"
    class="v-button"
  >
    <slot/>
  </button>
</template>

<script>
/**
 * @module Button
 */
export default {
  name: 'VButton',
  props: {
    color: {
      type: String,
      default: 'primary'
    },
    outlined: {
      type: Boolean,
      default: false
    }
  }
}
</script>
