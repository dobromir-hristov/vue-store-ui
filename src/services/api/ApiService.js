/* Here is where we could create our Api service, currently a mocked one */
const DELAY = 500

const delayedResolvePromise = (delay = DELAY) => new Promise(resolve => setTimeout(() => resolve(), delay))

const axiosInstance = {
  get: () => delayedResolvePromise(),
  post: () => delayedResolvePromise(),
  put: () => delayedResolvePromise(),
  delete: () => delayedResolvePromise()
}

export const ApiService = {
  get: axiosInstance.get,
  post: axiosInstance.post,
  put: axiosInstance.put,
  delete: axiosInstance.delete
}
