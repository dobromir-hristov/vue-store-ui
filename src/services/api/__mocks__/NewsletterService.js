export const NewsletterService = {
  subscribe: jest.fn(() => Promise.resolve()),
  unsubscribe: jest.fn(() => Promise.resolve())
}
