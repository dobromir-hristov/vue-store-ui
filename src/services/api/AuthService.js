import { ApiService } from './ApiService'

const GET_USER_ENDPOINT = '/user'
const LOGIN_ENDPOINT = '/login'

export const AuthService = {
  /**
   * Fetches the user by the token
   * @param {string} token
   * @return {Promise}
   */
  async fetchUser (token) {
    await ApiService.get(GET_USER_ENDPOINT, { token })
    return { id: 'some_id', name: 'John Doe', email: 'john@doe.com' }
  },

  /**
   * Logs in the user
   * @param {{ email: string, password:string }} credentials
   * @return {Promise}
   */
  async loginUser (credentials) {
    await ApiService.post(LOGIN_ENDPOINT, credentials)
    return { token: 'some_fake_token' }
  }
}
