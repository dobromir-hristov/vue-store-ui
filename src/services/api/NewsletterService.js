import { ApiService } from './ApiService'

const SUBSCRIBE_ENDPOINT = '/newsletter/subscribe'
const UNSUBSCRIBE_ENDPOINT = '/newsletter/unsubscribe'

export const NewsletterService = {
  subscribe (userId, newsletterId) {
    return ApiService.post(SUBSCRIBE_ENDPOINT, { user_id: userId, newsletter_id: newsletterId })
  },
  unsubscribe (userId, newsletterId) {
    return ApiService.post(UNSUBSCRIBE_ENDPOINT, { user_id: userId, newsletter_id: newsletterId })
  }
}
