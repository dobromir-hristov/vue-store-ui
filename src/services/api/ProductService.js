import { ApiService } from './ApiService'

const GET_PRODUCT_ENDPOINT = '/products'

export const ProductService = {
  fetchProduct (productId) {
    return ApiService.get(`${GET_PRODUCT_ENDPOINT}/${productId}`)
  }
}
