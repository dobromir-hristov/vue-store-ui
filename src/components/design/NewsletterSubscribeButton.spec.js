import VButton from './VButton'
import flushPromises from 'flush-promises'
import NewsletterSubscribeButton from './NewsletterSubscribeButton'
import { NewsletterService } from '@/services/api/NewsletterService'

jest.mock('@/services/api/NewsletterService')

/**
 * @param overrides
 * @param storeOverrides
 * @returns {{config, wrapper: Wrapper<V> | Wrapper<Vue> | *}}
 */
const createWrapper = (overrides, storeOverrides) => {
  const store = {
    user: {
      getters: {
        userId: () => 1,
        isLoggedIn: () => true
      }
    }
  }

  let config = _merge(createComponentMocks({
    store: _merge(store, storeOverrides)
  }), overrides)
  const wrapper = mount(NewsletterSubscribeButton, config)
  return {
    config,
    wrapper
  }
}

describe('NewsletterSubscribeButton', () => {
  it('renders the NewsletterSubscribeButton', () => {
    const { wrapper } = createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('disables if logged out', () => {
    const { wrapper } = createWrapper({}, {
      user: {
        getters: {
          isLoggedIn: () => false
        }
      }
    })
    expect(wrapper.find('button').attributes('disabled')).toBeTruthy()
  })

  it('toggles the user subscription', async () => {
    const { wrapper } = createWrapper()
    expect(wrapper.text()).toContain('Subscribe')
    wrapper.find(VButton).trigger('click')
    expect(wrapper.find(VButton).props('loading')).toBe(true)
    await flushPromises()

    expect(NewsletterService.subscribe).toHaveBeenCalledTimes(1)
    expect(wrapper.text()).toContain('Unsubscribe')
    wrapper.find(VButton).trigger('click')
    await flushPromises()
    expect(NewsletterService.unsubscribe).toHaveBeenCalledTimes(1)
    expect(wrapper.text()).toContain('Subscribe')
  })
})
