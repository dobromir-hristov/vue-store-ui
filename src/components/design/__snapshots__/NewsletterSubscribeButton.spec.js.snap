// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsletterSubscribeButton renders the NewsletterSubscribeButton 1`] = `<button class="v-button v-button--primary is-outlined v-button--size-large" type="alternative"><span class="v-button__text"><span>Subscribe</span></span></button>`;
